contracts/governance/KyberStaking.sol

```diff
@@ -0,0 +1,479 @@
+
+    // increase delegated stake for address that staker has delegated to (if it is not staker)
+    address representative = stakerPerEpochData[curEpoch + 1][staker].representative;
+    if (representative != staker) {
+      initDataIfNeeded(representative, curEpoch);
+      stakerPerEpochData[curEpoch + 1][representative].delegatedStake = stakerPerEpochData[
+        curEpoch + 1
+      ][representative]
+        .delegatedStake
+        .add(amount);
+      stakerLatestData[representative].delegatedStake = stakerLatestData[representative]
+        .delegatedStake
+        .add(amount);
+    }
+
+    emit Deposited(curEpoch, staker, amount);
+  }
+
+  /**
+   * @dev call to withdraw KNC from staking, it could affect voting point when calling withdrawHandlers handleWithdrawal
+   * @param amount amount of KNC to withdraw
+   */
+  function withdraw(uint256 amount) external override nonReentrant {
+    require(amount > 0, 'withdraw: amount is 0');
+
+    uint256 curEpoch = getCurrentEpochNumber();
+    address staker = msg.sender;
+
+    require(
+      stakerLatestData[staker].stake >= amount,
+      'withdraw: latest amount staked < withdrawal amount'
+    );
+
+    (bool success, ) =
+      address(this).call(
+        abi.encodeWithSignature(
+          'handleWithdrawal(address,uint256,uint256)',
+          staker,
+          amount,
+          curEpoch
+        )
+      );
+    if (!success) {
+      // Note: should catch this event to check if something went wrong
+      emit WithdrawDataUpdateFailed(curEpoch, staker, amount);
+    }
+
+    stakerLatestData[staker].stake = stakerLatestData[staker].stake.sub(amount);
+
+    // transfer KNC back to staker
+    require(kncToken.transfer(staker, amount), 'withdraw: can not transfer knc');
+    emit Withdraw(curEpoch, staker, amount);
+  }
+
+  /**
+   * @dev initialize data if needed, then return staker's data for current epoch
+   * @dev for safe, only allow calling this func from KyberDao address
+   * @param staker - staker's address to initialize and get data for
+   */
+  function initAndReturnStakerDataForCurrentEpoch(address staker)
+    external
+    override
+    returns (
+      uint256 stake,
+      uint256 delegatedStake,
+      address representative
+    )
+  {
+    uint256 curEpoch = getCurrentEpochNumber();
+    initDataIfNeeded(staker, curEpoch);
+
+    StakerData memory stakerData = stakerPerEpochData[curEpoch][staker];
+    stake = stakerData.stake;
+    delegatedStake = stakerData.delegatedStake;
+    representative = stakerData.representative;
+  }
+
+  /**
+   * @notice return raw data of a staker for an epoch
+   *         WARN: should be used only for initialized data
+   *          if data has not been initialized, it will return all 0
+   *          pool master shouldn't use this function to compute/distribute rewards of pool members
+   * @dev  in KyberDao contract, if staker wants to claim reward for past epoch,
```

Comment:
Please change the comments, some of them are related to the old one

Reply:
change as request